HARJA-1498 - Pysyvät muutokset jatkuu ja muutosten lähdekoodin jakaminen erillisiin tiedostoihin

solita-staskila opened the pull request:
## Muutosten hallinta - Pysyvät muutokset

* Paljon refaktoroitua koodia ja kyselyitä uudistettu
* Versionnin logiikkaa hiukan muunnettu eri suuntaan, kuin alunperin on suunniteltu (vain muuttuvat rivit nousevat versiossa, äitimuutoksessa tallessa aina korkein versionumero)
* Pysyvien muutosten tallennus usealle hoitokaudelle, kattavat testit
* Uuden pysyvän muutoksen luonti (alustava, testaamaton) 
* Kopiointi tuleville hoitovuosille (alustava, ei kunnolla testattu)
  * Ainakin yksi tiedossa oleva bugi 

* Muutoshistorian tietomallia laitettu uusiksi
  * Käytetään Inherits sijasta like-komennolla luotuja historiatauluja, triggereitä paranneltu 
  
* Korjailtu ongelmia testidatassa ja tehty sitä kattavammaksi
  * Nyt tuotu II:n MHU:n lisäksi Suomussalmen MHU, joka pidempään jatkuvana testiurakkana sopii hyvin muutosten testaamiseen. 


### Pilkottu näkymät näin:

```c++
harja.views.urakka 
└── muutokset
    ├── lomake
    │   ├── lomake_johto_hallinto.cljs // tyyppikohtaiset lomakkeet
    │   ├── lomake_pysyva.cljs     // tyyppikohtaiset lomakkeet
    │   ├── ...... ja muut              // tyyppikohtaiset lomakkeet
    │   └── muutoslomake.cljs  // lomake päätiedosto 
    ├── kirjatut_muutokset.cljs   // muutos-osio
    ├── lasketut_muutokset.cljs  // muutos-osio
    ├── rahavarausten_muutokset.cljs  // muutos-osio
    └── yhteiset.cljs  // yhteiset
```

```c++
harja.tiedot.urakka 
└── muutokset
    ├── kirjatut_muutokset_tiedot.cljs 
    ├── lasketut_muutokset_tiedot.cljs
    ├── rahavarausten_muutokset_tiedot.cljs
    └── yhteiset_tiedot.cljs
```

<img width="442" height="244" alt="image" src="https://github.com/user-attachments/assets/e65ed291-17e4-44f9-88c3-80d5cc14372b" />

<img width="325" height="206" alt="image" src="https://github.com/user-attachments/assets/f6c1d576-f592-4652-a6fe-fd803fce1aaa" />


### Muuta
* CI-putkessa cachetaan jatkossa Cypress-binäärit, jotta niitä ei tarvitse ladata joka kerta. Vähentää korruptoituneen latauksen mahdollisuutta
* Käytetään testeissa uusinta saatavilla olevaa testikantaa vanhan PostgreSQL 13 sijasta
  * Vanhoissa testiympäristöissä huomioitava vielä, että siellä ei voi hyödyntää Possu 13 versiota uudempia ominaisuuksia.
* Tuotu joukko kokoelma-apureita eri tarkoituksiin: https://github.com/finnishtransportagency/harja/blob/cb346b09284fdca1afcb3fd782a71215df55e786/src/cljc/harja/kokoelmat.cljc

jarnovayrynen committed:
Alusta pysyvän muutoksen lomaketta

-Lisää hoitokausivalinta ja oletusarvon asettaminen

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

jarnovayrynen committed:
Siirrä migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

jarnovayrynen committed:
Aloita pysyvän muutoksen tietokantahaku ja UI

jarnovayrynen committed:
Siirrä migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

jarnovayrynen committed:
Lisää testidataa ja SQL-kyselyä pysyvän muutoksen tiedoille

Mahdollista se että palvelu palauttaa muutoksen kaikkien hoitovuosien tiedot kerralla

jarnovayrynen committed:
Korjaa että vain pysyvä muutos saa pysyvän tiedot

jarnovayrynen committed:
Tuo solun-luokka mahdolliseksi gridiin myös kun ei muokata

jarnovayrynen committed:
Pysyvän muutoksen hv-valinta vain Voimassa alkaen -jälkeen

jarnovayrynen committed:
Siirrä migraationumeroa

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

jarnovayrynen committed:
Lisää toimintoja pysyvän muutoksen taulukkoon

jarnovayrynen committed:
Tee tavoitehinnan muutoksen komponentti

Lisää eventin placeholder

jarnovayrynen committed:
Lisää kyky tarkkailla tehtävämäärägridin muutoksia

jarnovayrynen committed:
Lisää toimenpidekoodi testeihin

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

jarnovayrynen committed:
Poista vanha muokkaus-gridiä varten ollut atomi

Ensimmäisessä vaiheessa pysyvän muutoksen taulukon toteutukseen oli ajateltu muokkaus-gridiä. Tämähän olisi syönyt sisäänsä atomin. Nyt kun toteutus muutettiin käyttämään tavan gridiä, poistetaan turhaksi jäänyt atomi.

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

K1ll3rF0x committed:
Nimeä migraatio uudelleen

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

solita-staskila committed:
Korjaa UI tyylejä Figman mukaiseksi

solita-staskila committed:
Merge branch 'HARJA-1369-a' into HARJA-1498_pysyva_muutos

solita-staskila committed:
Korjaa konfliktoivat importit

solita-staskila committed:
Merge branch 'HARJA-1369-a' into HARJA-1498_pysyva_muutos

solita-staskila committed:
Muutostyö lomake frontend & haaroita lomakkeen tyyppejä

K1ll3rF0x committed:
Lisää debug-lokitusta ja TODO-kommentteja

K1ll3rF0x committed:
Nimeä migraatio uudelleen

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498_pysyva_muutos

K1ll3rF0x committed:
Refaktoroi uuden pysyvän muutoksen tietojen hakua, lisää TODO
kommentteja

K1ll3rF0x committed:
Poista INHERITS-rakennet muutosten historian tietomallista
* Annetaan historiataulujen elää omaa elämäänsä hyödyntämällä LIKE
komentoa INHERITS sijasta

K1ll3rF0x committed:
Poista vanha harjadb-versio PR-testiputkesta, se on jäänne
pilvisiirtymästä

solita-staskila committed:
Pilko muutos välilehden osiot eri namespaceen

K1ll3rF0x committed:
Käytä uudempaa nodejs:ää ja cacheta cypress binääri

K1ll3rF0x commented on src/cljs/harja/views/urakka/muutos_nakyma.cljs:
```diff
@@ -622,38 +132,41 @@
      (when-not indeksikorjaus-vahvistettu? [yleiset/vihje "Indeksikorjaus vahvistetaan kustannussuunnitelmassa."])]))
 
 
+(defn muutosten-hallinta-header [e! {:keys [haku-kaynnissa?] :as app}]
```
Onhan tässä enemmänkin kuin header, hiukan hämäävä nimi.

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Lisää apureita kokoelmien (coll) käsittelyyn

K1ll3rF0x committed:
Nimeä komponentti uudelleen

solita-staskila committed:
Merge branch 'HARJA-1498-frontti' into HARJA-1503

K1ll3rF0x committed:
Ryhmittele muutos-tiedot event-handlereita uusiksi

solita-staskila committed:
Muutostyö lomaketyypin frontti

solita-staskila committed:
Korjaa päivitys ja versiointi kustannusvaikutukseen

solita-staskila committed:
Lisää validointi sarakkeeseen

solita-staskila committed:
Merge branch 'HARJA-1498-frontti' into HARJA-1503

solita-staskila committed:
Yritä korjata yksikkötesti

K1ll3rF0x committed:
Palastele muutosten tilanhallinta erillisiin tiedostoihin

K1ll3rF0x committed:
Merge branch 'HARJA-1498-frontti' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Korjaa require

K1ll3rF0x committed:
Siirrä migraatio 1403 ennen migraatio 1308

solita-staskila committed:
Yritä korjata fronttitesti

solita-staskila committed:
Yritä korjata fronttitesti

solita-staskila committed:
Kokeile hajooko testit

solita-staskila committed:
Kokeile hajooko testit

K1ll3rF0x committed:
Merge branch 'muutosten-historia-tietomallimuutos' into HARJA-1498-frontti

K1ll3rF0x committed:
Muuta migraation järjestystä

K1ll3rF0x committed:
Siirrä koodia migraatiosta V1_1310 R-migraatioon, lisää TODO, tee
korjauksia

K1ll3rF0x committed:
Säädä muutosten tietomallin indeksejä ja rajoitteita

K1ll3rF0x committed:
Palauta takaisin mhu_muutos -> lapsi versiointilogiikka, pura liian
tiukkoja constraintteja

K1ll3rF0x committed:
Tue usean kustannusvaikutuksen tallentamista muutoksen yhteydessä

K1ll3rF0x committed:
Korjaa constraintteja

K1ll3rF0x committed:
Parantele kulun lisäystä, lisää FIXME-kommentteja

K1ll3rF0x committed:
Lisää TODO

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Tee geneerisempi luo_mhu_muutoksia ja lisää muutoksia Suomussalmen
urakkaan

K1ll3rF0x committed:
Korjauksia muutosten testidataan ja testeihin

K1ll3rF0x committed:
Korjaa analytiikan testi

K1ll3rF0x committed:
Korjaa indeksivirhe

K1ll3rF0x committed:
Korjaa rivin poiston tapaus muutoshistorian triggereihin

K1ll3rF0x committed:
Pura versio-check joineissa, lisää puuttuva versio upserttiin

solita-staskila committed:
Siirrä johto-ja-hallintokorvausmuutokset-atom tuckin tilaan

K1ll3rF0x committed:
Hahmottele määrämuutosten tallennusta, muuta muuttujien nimiä

K1ll3rF0x committed:
Korjaile liitteiden lisäämisen ja poistamisen logiikkaa

K1ll3rF0x committed:
Refaktoroi pysyvän muutoksen lomake, jotta gridiä ei turhaan renderöidä
* Gridin turha uudelleenrenderöinti johtui siitä, että app tila
kuljetettiin ui form-1 tyyppisille komponenteille sellaisenaan ja render
funktio reagoi kaikkii app tilan muutoksiin. Tämä esti mm. gridin
vetolaatikoiden auki pysymisen lomakkeen tietojen syötön aikana.

K1ll3rF0x committed:
Päivitä pysyvän muutoksen lomakkeen tehtävät-ja-määrät tilanhallinta

K1ll3rF0x committed:
Lisää tilanhallintaa pysyvän muutoksen tehtävien ja määrien tietojen
keräykseen

K1ll3rF0x committed:
Tue pysyvien muutosten tilanhallinnassa usean hoitovuoden tietojen
editointia

K1ll3rF0x committed:
Korjaa muutos testidata, korjaa urakan viimeinen hoitovuosi

K1ll3rF0x committed:
Tallenna tehtä- ja määrämuutoksia ja kustannusvaikutuksia usealta
hoitovuodelta

K1ll3rF0x committed:
Korjaa pysyvien muutosten haun testi, testaa useampi hoitovuosi

K1ll3rF0x committed:
Korjaa pysyviin muutoksiin liittyviä kyselyitä, lisää pysyvän muutoksen
tallennukseen liittyviä testejä, korjaa vanhoja testejä

K1ll3rF0x committed:
Lisää Suomussalmen urakalle sopimuksen tehtävämäärät testidataan

K1ll3rF0x committed:
Korjaa bugi tehtavien määrien muokkauksen tilanhallinnassa

K1ll3rF0x committed:
Korjaa bugi yksikkötiedon haussa toimenpide-gridissä

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Toteuta tehtävän määrämuutoksen rivin poisto tilanhallintaan,
yksinkertaista tilanhallintaa apufunktioin

K1ll3rF0x committed:
Toteuta tehtävän määrämuutosrivin poisto backendiin

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Estä turhaan tehtävän määrämuutosrivin päivitys, jos määrä ei muutu

K1ll3rF0x committed:
Estä turhaan kustannusvaikutuksen päivitys, jos summa ei muutu

K1ll3rF0x committed:
Mahdollista usean tyhjen määrämuutosrivin luonti, filtteröi pois
negatiiviset tehtävä-id:t tallennuksen yhteydesä

K1ll3rF0x committed:
Estä saman tehtävän valinta useammalle riville

K1ll3rF0x committed:
Poista vain UI:ssa olemassaolevat rivit kokonaan, mikäli ne poistetaan

K1ll3rF0x committed:
Hahmottele pysyvien muutosten tuleville vuosille kopioiminen

K1ll3rF0x committed:
Lisää FIXME-kommentti tuleville vuosille kopiointiin

K1ll3rF0x committed:
Aja poistot ja lisäykset/päivitykset kahdella eri kierroksella

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Nimeä migraatiot uudelleen

K1ll3rF0x committed:
Tuleville vuosille kopioinnissa poista rivi kokonaan, jos poistettava
rivi on vain uusi rivi, jota ei ole vielä tallennettu tietokantaan

K1ll3rF0x committed:
Erota pysyvien muutosten pohjatiedon haku omaan funktioon

K1ll3rF0x committed:
Hahmottele pysyvien muutosten pohjatiedon haku tyhjälle lomakkeelle

K1ll3rF0x committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into HARJA-1498-frontti

K1ll3rF0x committed:
Vie pysyvän muutokset pohjatiedot tyhjälle muutoslomakkeelle

K1ll3rF0x committed:
Refaktoroi pysyvän muutoksen pohjatietojen hakua

K1ll3rF0x committed:
Resetoi hoitovuosi tarvittaessa voimassa-alkaen pvm muokatessa

K1ll3rF0x committed:
Korjaa bugi, jossa muutosta muokatessa valittiin mahdoton hoitovuosi
lomakkeelle

solita-staskila commented on src/cljs/harja/views/urakka/muutokset/lomake/muutoslomake.cljs:
```diff
@@ -0,0 +1,129 @@
+       (t-yhteiset/scrollaa-viimeksi-valitulle-riville)
+       (e! (t-yhteiset/->MuokkaaMuutosta nil)))
+    {:disabled tallennus-kesken?}]])
+
+(defn- alusta-lomakkeen-pohjatiedot [e! muutostyyppi valittu-hoitokausi rivi]
+  (log/debug "Haetaan lomakkeen pohjatiedot muutostyypille:" muutostyyppi)
+
+  (let [rivi (case muutostyyppi
+               "pysyva" (do
+                          (e! (t-kirjatut/->HaePysyvanMuutoksenPohjatiedotLomakkeelle))
+                          rivi)
+               ;; TODO: Tämä on jäänne muutosten ensimmäisen version ajalta, ja poikkeaa muusta logiikasta
+               ;;       Vaatisi refaktorointia, jotta logiikka lomakkeen alustamiselle olisi yhtenäisempi
+               "johto-ja-hallintokorvaus" (assoc rivi
+                                            :johto-ja-hallintokorvaukset
+                                            (t-yhteiset/johto-ja-hallintokorvausmuutoksen-rivit valittu-hoitokausi [])))]
+    rivi))
```
Pitäs jokin defaultti olla, ei muuten pysty tehdä muita lomakkeita 

```clojure
;; Default 
(do (e! (t-yhteiset/->MuokkaaMuutosta {})) 
    rivi)
```

Tää ilmeisesti vähän vaiheessa, heittää virhettä, jos yrittää tehdä uutta pysyvää muutosta
-> error: (e! (t-kirjatut/->HaePysyvanMuutoksenPohjatiedotLomakkeelle))

solita-staskila reviewed (commented):
Jaha, oma pullari, en voi itse approvaa.

Katsoin koodit läpi, ei nyt tässä vaiheessa oikeastaan mitään kummempia sanomisia. Siellä on sen verran omaakin välissä (mikä on tietenkin täydellistä koodia) 

Naputtelin käliä vähäsen, näyttää kannassa menevän kaikki suht oikein, siihen asti mitä on tehty. Hyvän näköisiä testejä. 

Kokoelmat saa mun puolesta myös tulla, on oikeastaan ollut tarvekin tuolle dissoc-in, kun clojuressa sitä ei ole 

Imaginary approved, kyllä tästä voidaan jatkaa, voi mergttää 

<img width="276" height="211" alt="image" src="https://github.com/user-attachments/assets/a4acd26a-7962-4c70-a8ad-6c526130b8c2" />

K1ll3rF0x committed:
Lisää default arvo pohjatietojen case-kutsuun

K1ll3rF0x committed:
Poista automaattinen hoitovuoden valinta uutta pysyvää muutosta tehdessä

K1ll3rF0x committed:
Merge branch 'develop' into HARJA-1498-frontti

K1ll3rF0x committed:
Poista vanhentunut TODO

K1ll3rF0x committed:
Merge branch 'HARJA-1498-frontti' of github.com:finnishtransportagency/harja into HARJA-1498-frontti
